### ISSUE TITLE
DynamoDB:UpdateItem ReturnValuesOnConditionCheckFailure flag does not work

### ISSUE THREAD

**Issue by drshriveer** [aws/aws-sdk-go-v2#2600]
### Acknowledgements

- [X] I have searched (https://github.com/aws/aws-sdk/issues?q=is%3Aissue) for past instances of this issue
- [X] I have verified all of my SDK modules are up-to-date (you can perform a bulk update with `go get -u github.com/aws/aws-sdk-go-v2/...`)

### Describe the bug

ConditionalCheckFailures on an update attempt with the `ReturnValuesOnConditionCheckFailure` set to `types.ReturnValuesOnConditionCheckFailureAllOld` does not return (or populate) the item (`*types.ConditionalCheckFailedException.Item`) as required by the API contract when running against DynamoDB**Local**. 

This almost certainly is the fault of DDBLocal rather than the go SDK. Unfortunately, there is no specific place to file a DynamoDBLocal bug that I am aware of.

### Expected Behavior

The error returned, which is a `*types.ConditionalCheckFailedException` should have the `Item` property filled with the Item's attribute map as it was at the time of the failed condition check. 
The expected behavior is what happens when running against production DynamoDB.

### Current Behavior

`types.ConditionalCheckFailedException.Item` is always empty when running against DDBLocal.

### Reproduction Steps

Here's a sample test that repos 100% of the time when running against DDBLocal. 

```
func TestTest(t *testing.T) {
	client := makeDDBLocalClient(t)
	ctx := context.TODO()
	tableName := "broken-api-contract"
	pk := "update-doesnt-work"

	assert.Eventually(t, func() bool {
		_, err := client.CreateTable(ctx, &dynamodb.CreateTableInput{
			TableName: aws.String(tableName),
			AttributeDefinitions: []types.AttributeDefinition{
				{AttributeName: aws.String(pk), AttributeType: types.ScalarAttributeTypeS},
			},
			KeySchema: []types.KeySchemaElement{
				{KeyType: types.KeyTypeHash, AttributeName: aws.String(pk)},
			},
			BillingMode: types.BillingModePayPerRequest,
		})
		return err == nil
	}, 10*time.Second, 100*time.Millisecond)

	propThatExists := "P"
	itemKey := "item-demonstrating-untestable-feature"

	_, err := client.PutItem(ctx, &dynamodb.PutItemInput{
		Item: map[string]types.AttributeValue{
			pk:             &types.AttributeValueMemberS{Value: itemKey},
			propThatExists: &types.AttributeValueMemberS{Value: "oh, I exist."},
		},
		TableName: aws.String(tableName),
	})

	require.NoError(t, err)

	_, err = client.UpdateItem(ctx, &dynamodb.UpdateItemInput{
		TableName: aws.String(tableName),
		Key: map[string]types.AttributeValue{
			pk: &types.AttributeValueMemberS{Value: itemKey},
		},
		ConditionExpression: aws.String("attribute_not_exists(" + propThatExists + ")"),
		UpdateExpression:    aws.String("SET " + propThatExists + " = :val"),
		ExpressionAttributeValues: map[string]types.AttributeValue{
			":val": &types.AttributeValueMemberS{Value: "I've been updated."},
		},
	})
	require.Error(t, err)
	var underlyingError *types.ConditionalCheckFailedException
	require.ErrorAs(t, err, &underlyingError)
	assert.NotEmpty(t, underlyingError.Item)
}
```

### Possible Solution

DDBLocal should handle the `ReturnValuesOnConditionCheckFailure` flag appropriately.

### AWS Go SDK V2 Module Versions Used

Using DDBLocal DockerImage: `amazon/dynamodb-local:2.1.0`

Not particularly relevant but AWS SDK versions:

```
	github.com/aws/aws-sdk-go-v2 v1.26.1
	github.com/aws/aws-sdk-go-v2/config v1.27.10
	github.com/aws/aws-sdk-go-v2/credentials v1.17.10
	github.com/aws/aws-sdk-go-v2/feature/dynamodb/expression v1.7.13
	github.com/aws/aws-sdk-go-v2/service/dynamodb v1.31.1
	github.com/aws/smithy-go v1.20.2
```

### Compiler and Version used

go version go1.22.1 darwin/arm64

### Operating System and version

macOS 13.5.2

**Comment by RanVaknin** [comment]
Hey @drshriveer , thanks for letting us know! You mentioned that this is happening on DDBLocal, does this mean that you tested it against the remote dynamo server and the response does contain the `ReturnValuesOnConditionCheckFailure` info as expected?

I know this feature is relatively recent. There's a chance you might need to update your DynamoDB local image to the latest version. I can see in the ECR public image gallery that the [latest version](https://gallery.ecr.aws/aws-dynamodb-local/aws-dynamodb-local) is 2.3.0


If even after updating your image this still doesn't work, I can help upstream a request to the DynamoDB local team on your behalf.

Thanks,
Ran~

**Comment by drshriveer** [comment]
My bad, I was running on an older version of the image, however the most recent image `2.3.0` also fails.

**Comment by RanVaknin** [comment]
HI @drshriveer ,

Thanks for the update. I tested this myself using the latest docker image:

```console
$ docker run -p 8000:8000 amazon/dynamodb-local:latest

Initializing DynamoDB Local with the following configuration:
Port:   8000
InMemory:       true
Version:        2.3.0
DbPath: null
SharedDb:       false
shouldDelayTransientStatuses:   false
CorsParams:     null
```

Create the table on our local dynamo server
```console
$ aws dynamodb create-table \
    --table-name TestTable \
    --attribute-definitions AttributeName=pk,AttributeType=S \
    --key-schema AttributeName=pk,KeyType=HASH \
    --billing-mode PAY_PER_REQUEST \
    --endpoint-url http://localhost:8000
```

Add an item:
```console
$ aws dynamodb put-item \
    --table-name TestTable \
    --item '{"pk": {"S": "testKey"}, "attr": {"S": "testValue"}}' \
    --endpoint-url http://localhost:8000
```

Run the Go SDK code against our table:

```go
package main

import (
	"context"
	"errors"
	"fmt"
	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/dynamodb"
	"github.com/aws/aws-sdk-go-v2/service/dynamodb/types"
)

func main() {
	cfg, err := config.LoadDefaultConfig(context.Background(),
		config.WithRegion("local-env"),
		config.WithEndpointResolverWithOptions(aws.EndpointResolverWithOptionsFunc(func(service, region string, options ...interface{}) (aws.Endpoint, error) {
			return aws.Endpoint{
				URL:           "http://localhost:8000",
				SigningRegion: "local-env",
			}, nil
		})),
	)
	if err != nil {
		panic(err)
	}

	client := dynamodb.NewFromConfig(cfg)

	input := &dynamodb.UpdateItemInput{
		TableName: aws.String("TestTable"),
		Key: map[string]types.AttributeValue{
			"pk": &types.AttributeValueMemberS{Value: "testKey"},
		},
		UpdateExpression: aws.String("SET #v = :value"),
		ExpressionAttributeNames: map[string]string{
			"#v": "value",
		},
		ExpressionAttributeValues: map[string]types.AttributeValue{
			":value": &types.AttributeValueMemberS{Value: "testValue"},
		},
		ConditionExpression:                 aws.String("attribute_not_exists(pk)"),
		ReturnValuesOnConditionCheckFailure: types.ReturnValuesOnConditionCheckFailureAllOld,
	}

	_, err = client.UpdateItem(context.TODO(), input)
	if err != nil {
		var cfe *types.ConditionalCheckFailedException
		if errors.As(err, &cfe) {
			fmt.Printf("Failed item details: %#v\n", cfe)
		} else {
			fmt.Println("Error:", err.Error())
		}
		return
	}
}
```

I'm able to see the failed updated values:
```
Failed item details: &types.ConditionalCheckFailedException{Message:(*string)(nil), ErrorCodeOverride:(*string)(nil), Item:map[string]types.AttributeValue{"attr":(*types.AttributeValueMemberS)(0x1400000e528), "pk":(*types.AttributeValueMemberS)(0x1400000e510)}, noSmithyDocumentSerde:document.NoSerde{}}
```

I'm not entirely sure why you are running into this problem, it says your SDK version is newer. Can you enable the response logs?

Can you also confirm the following:
> You mentioned that this is happening on DDBLocal, does this mean that you tested it against the remote dynamo server and the response does contain the ReturnValuesOnConditionCheckFailure info as expected?


Thanks,
Ran~

**Comment by drshriveer** [comment]
Thanks @RanVaknin! We figured it out. Indeed, the issue was fixed in the newer docker image `2.3.0`. The problem is that when I modified my original code to be a workable example (to share in this issue) I failed to copy over the _rather critical_ line `ReturnValuesOnConditionCheckFailure: types.ReturnValuesOnConditionCheckFailureAllOld`. And sadly, when I went to go re-check against 2.3.0, I copied my own invalid example code, which expectedly failed because the flag wasn't set. 🤦‍♂️ . 

My apologies and thanks for your time. I will close this as it is resolved.
